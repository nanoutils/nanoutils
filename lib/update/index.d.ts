import {
  CurriedFunction1,
  CurriedFunction2,
  Index,
  MixedArray
} from '../../misc/types/fn'

declare function update<T1, T2>(index: Index): CurriedFunction2<T2, Array<T1>, MixedArray<T1, T2>>
declare function update<T1, T2>(index: Index, value: T2): CurriedFunction1<Array<T1>, MixedArray<T1, T2>>
declare function update<T1, T2>(index: Index, value: T2, array: Array<T1>): MixedArray<T1, T2>

export default update