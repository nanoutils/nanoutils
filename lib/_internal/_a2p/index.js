export default (a) =>
  Array.isArray(a)
    ? a.join('.')
    : typeof a === 'string' || typeof a === 'number'
    ? '' + a
    : ''
