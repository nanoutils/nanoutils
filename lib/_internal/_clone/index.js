export function shallowCloneObjectAndArrays(obj) {
  var res = {}
  if (typeof obj !== 'object') {
    return res
  }
  if (Array.isArray(obj)) {
    return obj.slice()
  }
  for (var k in obj) {
    res[k] = obj[k]
  }
  return res
}
