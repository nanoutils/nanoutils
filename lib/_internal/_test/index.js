/**
 * Check for accepting number of arguments for given function
 * @param {function} f - function to check
 * @param {number} n - number of accepting arguments
 * @param {array} args - example of arguments to check
 */
export function expectNumberOfArgs(f, n, args) {
  let k = n
  while (k > 0) {
    expect(typeof f).toBe('function')
    f = f(args[n - k--])
  }
  expect(typeof f).not.toBe('function')
}

export function expectExactNumberOfArgs(f, args) {
  expect(typeof f.apply(f, args)).not.toBe('function')
}
