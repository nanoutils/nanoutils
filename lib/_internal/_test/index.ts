import { expect } from "earljs";

/**
 * Check for accepting number of arguments for given function
 * @param {function} f - function to check
 * @param {number} n - number of accepting arguments
 * @param {array} args - example of arguments to check
 */
export function expectNumberOfArgs<Args extends unknown>(
  f: (args: Args) => unknown,
  n: number,
  args: Args[]
) {
  let k = n;
  let g: (args: Args) => unknown = f;
  while (k > 0) {
    expect(g).toBeA(Function);
    g = g(args[n - k--]) as (args: Args) => unknown;
  }
  expect(g).not.toBeA(Function);
}

export function expectExactNumberOfArgs<Args extends unknown>(
  f: (args: Args) => unknown,
  args: Args[]
) {
  expect(f.apply(f, args)).not.toBeA(Function);
}
