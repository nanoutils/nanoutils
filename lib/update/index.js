import _curry3 from '../_internal/_curry3'
import update from './update'

export default _curry3(update)
