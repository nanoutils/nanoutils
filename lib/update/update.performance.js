const {
  generators: { getArrays },
  times: { TIMES_1M },
  types: { TYPE_1M }
} = require('../_internal/helpers/performance')

module.exports = function() {
  const chBefore = 'a'
  const chAfter = 'b'
  const [big1, big2, big3] = getArrays(TIMES_1M, () => chBefore)

  return {
    type: TYPE_1M,
    argss: [
      [big1.length / 2, chAfter, big1],
      [big2.length / 2, chAfter, big2],
      [big3.length / 2, chAfter, big3]
    ]
  }
}
