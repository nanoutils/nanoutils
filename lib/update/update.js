export default function update(index, value, array) {
  if (process.env.NODE_ENV !== 'production') {
    require('../_internal/_error').checkTypes('update', {
      index: [index, ['number'], typeof index !== 'number'],
      value: [value, ['any'], false],
      array: [array, ['array'], !Array.isArray(array)]
    })
  }
  var result = array.slice()
  if (index >= 0 && index < result.length) {
    result[index] = value
  }
  return result
}
