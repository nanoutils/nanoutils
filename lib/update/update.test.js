import { expectNumberOfArgs } from '../_internal/_test'
import update from '.'

test('it accepts exact 3 arguments', () => {
  expectNumberOfArgs(
    update,
    3,
    [2, 10, [0, 1, 2]]
  )
})

test('updates value in list at the given index', () => {
  const list = [1, 2, 3, 4, 5]

  expect(update(1, 10, list)).toEqual([1, 10, 3, 4, 5])
})

test('does not mutate original array', () => {
  const list = [1, 2]
  const result = update(1, 10, list)

  expect(result).toEqual([1, 10])
  expect(list).toEqual([1, 2])
  expect(result !== list).toBeTruthy()
})

test('checks dev errors', () => {
  expect(() => update('1', 10, [1, 2, 3])).toThrowErrorMatchingSnapshot()
  expect(update(1, false, [1, 2, 3])).toEqual([1, false, 3])
  expect(() => update(1, 10, '123')).toThrowErrorMatchingSnapshot()
})
