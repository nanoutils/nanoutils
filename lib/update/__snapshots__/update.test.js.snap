// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks dev errors 1`] = `"[Nanoutils] update(index, *, *) accepts a number, string given"`;

exports[`checks dev errors 2`] = `"[Nanoutils] update(*, *, array) accepts an array, string given"`;
